core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -633,18 +837,27 @@ class ConnectionQuotasTest {
                                 listenerName: ListenerName,
                                 address: InetAddress,
                                 numConnections: Long,
-                                timeIntervalMs: Long) : Unit = {
+                                timeIntervalMs: Long,
+                                expectIpThrottle: Boolean): Boolean = {
     var nextSendTime = System.currentTimeMillis + timeIntervalMs
+    var ipThrottled = false
     for (_ <- 0L until numConnections) {
       // this method may block if broker-wide or listener limit on the number of connections is reached
-      connectionQuotas.inc(listenerName, address, blockedPercentMeters(listenerName.value))
-
+      try {
+        connectionQuotas.inc(listenerName, address, blockedPercentMeters(listenerName.value))
+      } catch {
+        case e: ConnectionThrottledException =>
+          if (!expectIpThrottle)
+            throw e
+          ipThrottled = true
```

Comment:
When we hit this line, should we just stopped trying more connections? `ipThrottled` will remain true regardless.

Reply:
a lot of the tests in `ConnectionQuotasTest` are using system time and are looking to reach a connection rate of N/s. so it may be the case that `ipThrottled` will remain true, but we want to verify that we can hit a target connection rate, even if connections in the middle of the interval get throttled.

Reply:
I took a closer look at some of the IP throttling tests, and I think they are a little too coarse-grained with respect to testing the throttle rate. I updated the IP throttling tests that don't expect to block to instead use `acceptConnectionsAndVerifyRate` with `MockTime`.

I think it should also be possible to update the broker/listener quota tests in `ConnectionQuotasTest` and `DynamicConnectionQuotaTest` to use mock time by extending the `Time` interface, but I would prefer to do that in a follow-up PR.

Reply:
Thanks. I filed https://issues.apache.org/jira/browse/KAFKA-10744 for the clean up/conversion.